rollouts/controllers/remoterootsync_controller_test.go

```diff
@@ -0,0 +1,114 @@
+			},
+		},
+		"with additional labels and annotations": {
+			rrsInput: `apiVersion: gitops.kpt.dev/v1alpha1
+kind: RemoteRootSync
+metadata:
+  name: my-remote-root-sync
+spec:
+  template:
+    metadata:
+      labels:
+       foo: bar
+      annotations:
+       abc: def
+       efg: hij`,
+			expectedActions: []clientgofake.Action{
+				clientgofake.NewPatchAction(rootSyncGVR,
+					"config-management-system",
+					"my-remote-root-sync",
+					"application/apply-patch+yaml",
+					[]byte(`{"apiVersion":"configsync.gke.io/v1beta1","kind":"RootSync","metadata":{"annotations":{"abc":"def","efg":"hij"},"labels":{"foo":"bar","gitops.kpt.dev/remoterootsync-name":"my-remote-root-sync","gitops.kpt.dev/remoterootsync-namespace":""},"name":"my-remote-root-sync","namespace":"config-management-system"}}`)),
+			},
+		},
+		"with source format": {
+			rrsInput: `apiVersion: gitops.kpt.dev/v1alpha1
+kind: RemoteRootSync
+metadata:
+  name: my-remote-root-sync
+spec:
+  template:
+    spec:
+      sourceFormat: unstructured`,
+			expectedActions: []clientgofake.Action{
+				clientgofake.NewPatchAction(rootSyncGVR,
+					"config-management-system",
+					"my-remote-root-sync",
+					"application/apply-patch+yaml",
+					[]byte(`{"apiVersion":"configsync.gke.io/v1beta1","kind":"RootSync","metadata":{"labels":{"gitops.kpt.dev/remoterootsync-name":"my-remote-root-sync","gitops.kpt.dev/remoterootsync-namespace":""},"name":"my-remote-root-sync","namespace":"config-management-system"},"spec":{"sourceFormat":"unstructured"}}`)),
+			},
+		},
+		"with git info": {
+			rrsInput: `apiVersion: gitops.kpt.dev/v1alpha1
+kind: RemoteRootSync
+metadata:
+  name: my-remote-root-sync
+spec:
+  template:
+    spec:
+      sourceFormat: unstructured
+      git:
+        repo: blueprints
+        branch: main
+        dir: namespaces`,
+			expectedActions: []clientgofake.Action{
+				clientgofake.NewPatchAction(rootSyncGVR,
+					"config-management-system",
+					"my-remote-root-sync",
+					"application/apply-patch+yaml",
+					[]byte(`{"apiVersion":"configsync.gke.io/v1beta1","kind":"RootSync","metadata":{"labels":{"gitops.kpt.dev/remoterootsync-name":"my-remote-root-sync","gitops.kpt.dev/remoterootsync-namespace":""},"name":"my-remote-root-sync","namespace":"config-management-system"},"spec":{"git":{"auth":"","branch":"main","dir":"namespaces","period":"0s","repo":"blueprints","secretRef":{}},"sourceFormat":"unstructured"}}`)),
+			},
+		},
+	}
+
+	for tn, tc := range testCases {
+		t.Run(tn, func(t *testing.T) {
+			var rrs gitopsv1alpha1.RemoteRootSync
+			require.NoError(t, yaml.Unmarshal([]byte(tc.rrsInput), &rrs))
+
+			client := dynamicfake.NewSimpleDynamicClient(scheme, &rrs)
+			(&RemoteRootSyncReconciler{}).patchRootSync(context.Background(), client, rrs.Name, &rrs)
+
+			require.Equal(t, tc.expectedActions, client.Actions())
```

Comment:
I am still thinking if we should just have unit test for `BuildObjectsToApply` and that will give the same effect. Testing the dynamic client part is not the intent here and that can only be tested in an e2e more effectively.

The e2e test for the behavior is, for a given input RRS, corresponding RS in target cluster should exist. Unit test version of that will be for a given RRS, BuildObjectsToApply(RRS) == expected RS.

WDYT ?

Reply:
SGTM! I will update this to test BuildObjectsToApply instead

Reply:
Updated this PR to test `BuildObjectsToApply`. I also added tests for two other helper functions that produce important inputs to `BuildObjectsToApply`.